### DISCUSSION TITLE
Discounts passed as an array with individual line id targets do not observe combinations

### DISCUSSION THREAD

**Discussion by limoniapps** [Shopify/function-examples#576] [Feedback: Discounts APIs]
Assume a cart contains two line items:
- 1x product variant A: $1500
- 1x product variant B: $500

Offer $100 off on each of those two products using a discount function:
- product discount
- fixed amount
- allocation method: each

Assume the function uses line ids to target the lines that must be discounted:
For example:
`{
  "discountApplicationStrategy": "ALL",
  "discounts": [
    {
      "message": "$100 OFF",
      "targets": [
        {
          "cartLine": {
            "id": "gid://shopify/CartLine/0",
            "quantity": 1
          }
        }
      ],
      "value": {
        "fixedAmount": {
          "amount": "100",
          "appliesToEachItem": true
        }
      }
    },
    {
      "message": "$100 OFF",
      "targets": [
        {
          "cartLine": {
            "id": "gid://shopify/CartLine/1",
            "quantity": 1
          }
        }
      ],
      "value": {
        "fixedAmount": {
          "amount": "100",
          "appliesToEachItem": true
        }
      }
    }
  ]
}`

At checkout both lines receive the correct discount.
Now add a Shopify discount code (ABC10) at checkout:
- product discount
- 10 percent
- targets the same two variants
- combines with product discounts

Expected result:
- Variant A: the 10% applies since it provides a better discount ($150 beats $100)
- Variant B: the $100 discount stands since the 10% is lower ($50)

Actual result:
The discount code is not applied "ABC10 couldn't be used with your existing discounts."

**Comment by adamjford** [comment]
Hey there! 👋 Developer from the Discounts team here. This is definitely a bug, considering the equivalent Function result that uses two targets within the same discount works as expected. We're looking into it. Thanks for the report!

**Comment by limoniapps** [comment]
Thanks @adamjford, I appreciate the fast turnaround!

**Comment by MacMaki** [comment]
### Possible Workarounds

The bug only arises when a product Function discount that has multiple proposals being applied is attempting to combine with another product discount that targets the same cart line(s). If the same discount value and message is being applied by each proposal, one possible workaround would be to combine the two proposals. For example, for the above FunctionRunResult, applying the workaround would look like this:

```
{
  "discountApplicationStrategy": "FIRST",
  "discounts": [
    {
      "message": "$100 OFF",
      "targets": [
        {
          "cartLine": {
            "id": "gid://shopify/CartLine/0",
            "quantity": 1
          }
        },
        {
          "cartLine": {
            "id": "gid://shopify/CartLine/1",
            "quantity": 1
          }
        }
      ],
      "value": {
        "fixedAmount": {
          "amount": "100",
          "appliesToEachItem": true
        }
      }
    }
  ]
}
```

**Comment by limoniapps** [comment]
Hi @adamjford Is there an update on this issue?

**Comment by MacMaki** [comment]
Follow up question to above suggested workaround:

When we say:

> “one possible workaround would be to combine the two proposals. For example, for the above FunctionRunResult, applying the workaround would look like this:”

 
What do we mean by “combine.” Obviously we’re gunning for the response to be the expected result:

Expected result:

Variant A: the 10% applies since it provides a better discount ($150 beats $100)
Variant B: the $100 discount stands since the 10% is lower ($50)
 
What we are not looking for this to do is for the 10% to stack or combine with the $100 off which would be double dipping 2 discounts.

Can we confirm that the workaround suggested would function as suggested here? @adamjford

**Reply by adamjford** [reply]
Yeah, "combine" was a loaded word to use there. Apologies!

What I mean is that the following two product discount `FunctionRunResult`s are equivalent in every way except for the bug in question:

Two proposal version (affected by bug):
```json
{
  "discountApplicationStrategy": "ALL",
  "discounts": [
    {
      "message": "$100 OFF",
      "targets": [
        {
          "cartLine": {
            "id": "gid://shopify/CartLine/0",
            "quantity": 1
          }
        }
      ],
      "value": {
        "fixedAmount": {
          "amount": "100",
          "appliesToEachItem": true
        }
      }
    },
    {
      "message": "$100 OFF",
      "targets": [
        {
          "cartLine": {
            "id": "gid://shopify/CartLine/1",
            "quantity": 1
          }
        }
      ],
      "value": {
        "fixedAmount": {
          "amount": "100",
          "appliesToEachItem": true
        }
      }
    }
  ]
}
```

One proposal version (not affected by bug):
```json
{
  "discountApplicationStrategy": "FIRST", # or MAXIMUM or ALL -- any would be equivalent
  "discounts": [
    {
      "message": "$100 OFF",
      "targets": [
        {
          "cartLine": {
            "id": "gid://shopify/CartLine/0",
            "quantity": 1
          }
        },
        {
          "cartLine": {
            "id": "gid://shopify/CartLine/1",
            "quantity": 1
          }
        }
      ],
      "value": {
        "fixedAmount": {
          "amount": "100",
          "appliesToEachItem": true
        }
      }
    }
  ]
}
```

These two produce product discounts that combine with other discounts in exactly the same way (except for, again, the bug in question). So, if you can change your discount Function to output a single proposal with multiple targets instead of multiple proposals with single targets (which, is, of course, only possible if they have the same value and message), that would avoid the bug.

**Reply by MacMaki** [reply]
@limoniapps

**Reply by limoniapps** [reply]
Hi @adamjford and @Maclainemaki 

We will implement the proposed workaround for the limited scenarios where it applies.
Note that we still need the underlying issue to be fixed by Shopify since this workaround has limitations and problems:
- Limitations: as Adam pointed out, the proposed workaround only works if the discount applied to both lines is the same (in this case $100 each), but cannot be applied if the discount is different for each line (e.g. $100 on line 1 and $150 on line 2).
- Problems: this workaround cannot be used for percentage discounts because, when multiple lines are grouped inside a discount, the discount is calculated based on the total value of all lines which leads to incorrect rounding for individual line items

**Reply by adamjford** [reply]
I fully agree with those limitations, and it's still a bug we're looking to fix for sure!

**Reply by limoniapps** [reply]
Hi @adamjford Is there an update on this?

**Comment by limoniapps** [comment]
Hi @adamjford Do we have an update on this? Is there a fix for this on the roadmap. We have more customers who are running into this scenario.
